File: cpu/instructions.py

"""
arg1 must be a register
arg2 must be a register
arg3 can be either a register or an immediate value
"""
ARITH_OPCODES = {
    'ADD' : lambda x, y: x + y,
    'SUB': lambda x, y: x - y,
    'MUL': lambda x, y: x * y,
    'DIV': lambda x, y: x // y,
    'MOD': lambda x,y : x % y,
    'AND': lambda x, y: x & y,
    'OR': lambda x, y : x | y,
    'SHL': lambda x, y: x << y,
    'SHR': lambda x, y: x >> y
    }

DATA_OPCODES = {
    # store in memory
    # arg1 must be a register, arg2 can be either register or immediate
    # MEM[arg1] = arg2
    'STR': None,  

    # load from memory
    # arg1 must be a register 
    # args2 can be either register or immediate
    # REG['arg1'] = MEM['arg2']
    'LOAD': None, # load from memory

    # move value to register  
    # arg1 must be a register
    # arg2 can be either a register or an immediate
    # REG[arg1] = arg2
    'MOV': None,

    # PUSH Rx: MEM[SP] = Rx, SP-=1
    'PUSH': None,
    # POP Rx: Rx = MEM[SP], SP+=1
    'POP': None,
}

TEST_OPCODES = {
    'TSTE': lambda x, y : int(x == y),
    'TSTG': lambda x, y: int(x > y),
    'TSTL': lambda x, y: int(x < y),
}
# instructions that affect IR (excluding RET)
FLOW_OPCODES = {
    'JUMP': lambda val : True,
    'TJMP': lambda val : val ,
    'FJMP': lambda val : not val,
    'CALL': lambda val : True
}

SPECIAL_OPCODES = {
    'HALT' : None,
    'RET' : None
}


def to_str(instruction):
    opcode = instruction['op']
    if opcode in ARITH_OPCODES:
        dst = instruction['dst']
        arg1 = instruction['arg1']
        arg2 = instruction['arg2']
        return f'{opcode} {dst} {arg1} {arg2}' 
    if opcode in DATA_OPCODES:
        if opcode == 'PUSH':
            return f"{opcode} {instruction['src']}"
        if opcode == 'POP':
            return f"{opcode} {instruction['dst']}"
        src = instruction['src']
        dst = instruction['dst']
        return f'{opcode} {dst} {src}'
    if opcode in TEST_OPCODES:
        arg1 = instruction['arg1']
        arg2 = instruction['arg2']
        return f'{opcode} {arg1} {arg2}'
    if opcode in FLOW_OPCODES or opcode in SPECIAL_OPCODES:
        return f'{opcode}'

def from_str(s):
    def maybe_cast_arg(arg):
        "this will try to cast arg to int, otherwise assumes its a register"
        try:
            return int(arg)
        except ValueError:
            return arg

    args = s.split()
    opcode = args[0]
    if opcode in ARITH_OPCODES:
        dst = args[1]
        arg1 = args[2] 
        arg2 = args[3] 
        return {'op': opcode, 'dst': dst, 'arg1': arg1, 'arg2': maybe_cast_arg(arg2)}
    if opcode in DATA_OPCODES:
        if opcode == 'PUSH':
            return {'op': opcode, 'src': maybe_cast_arg(args[1])}
        if opcode == 'POP':
            return {'op': opcode, 'dst': args[1]}
        dst = args[1]
        src = args[2] 
        return {'op': opcode, 'dst': dst, 'src': maybe_cast_arg(src)}
    if opcode in TEST_OPCODES:
        arg1 = args[1] 
        arg2 = args[2] 
        return {'op': opcode, 'arg1': arg1, 'arg2': maybe_cast_arg(arg2)}
    if opcode in FLOW_OPCODES:
        return {'op': opcode, 'offset': args[1]}
    if opcode in SPECIAL_OPCODES:
        return {'op': opcode}
    raise Exception('INVALID OPCODE:{}'.format(opcode))

File: src/instructions.rs

#[derive(Debug)]
enum Register{
    R1,
    R2,
    R3,
    R4,
    SP,
    BP,
    IR,
    ZR,
}

#[derive(Debug)]
enum BinArithOp{
    ADD,
    SUB,
    MUL,
    DIV,
    MOD,
    AND,
    OR,
    SHL,
    SHR,
    XOR,
}

impl BinArithOp{
    fn eval(&self, x: i32, y: i32) -> i32{
        match &self{
            BinArithOp::ADD => x + y,
            BinArithOp::SUB => x - y,
            BinArithOp::MUL => x * y,
            BinArithOp::DIV => x / y,
            BinArithOp::MOD => x % y,
            BinArithOp::AND => x & y,
            BinArithOp::OR => x | y,
            BinArithOp::SHL => x << y,
            BinArithOp::SHR => x >> y,
            BinArithOp::XOR => x ^ y,
        }
    }
}

#[derive(Debug)]
enum UnaryArithOp{
    NEG,
}

impl UnaryArithOp{
    fn eval(&self, x: i32) -> i32{
        match &self{
            UnaryArithOp::NEG => -x,
        }
    }
}

#[derive(Debug)]
enum DataOp{
    LOAD,
    STR,
    MOV,
}

#[derive(Debug)]
enum StackOp{
    PUSH,
    POP,
}

#[derive(Debug)]
enum TestOp{
    TSTE,
    TSTN,
    TSTG,
    TSTL,
}

impl TestOp{
    fn test(&self, arg1 : i32, arg2: i32) -> bool{
        match &self{
            TestOp::TSTE => arg1 == arg2,
            TestOp::TSTN => arg1 != arg2,
            TestOp::TSTG => arg1 > arg2,
            TestOp::TSTL => arg1 < arg2,
        }
    }
}

#[derive(Debug)]
enum FlowOp{
    JUMP,
    TJMP,
    FJMP,
    CALL,
}

impl FlowOp{
    fn should_take(&self, arg : i32) -> bool{
        match &self{
            FlowOp::JUMP => true,
            FlowOp::TJMP => arg != 0,
            FlowOp::FJMP => arg == 0,
            FlowOp::CALL => true,
        }

    }
}

#[derive(Debug)]
enum OtherOp{
    HALT,
    RET,
}

#[derive(Debug)]
enum RegOrImm{
    Reg(Register),
    Val(i32),
}

#[derive(Debug)]
enum Instruction{
    UnaryArith {op: UnaryArithOp, arg: Register},
    BinArith {op : BinArithOp, dst: Register, arg1: Register, arg2: RegOrImm},
    Data {op: DataOp, dst: Register, src: RegOrImm},
    Stack {op: StackOp, dst: Register},
    Test {op: TestOp, arg1: Register, arg2: RegOrImm},
    Flow {op: FlowOp, offset: i32},
    Other {op: OtherOp},
}

impl Instruction{
    fn to_str(&self) -> String{
        match &self{
            Instruction::UnaryArith{op, arg} => format!("{:?} {:?}", op, arg),
            Instruction::BinArith{op, dst, arg1, arg2} => format!("{:?} {:?} {:?} {:?}", op,  dst, arg1, arg2),
            Instruction::Data{op, dst, src} => format!("{:?} {:?} {:?}", op, dst, src),
            Instruction::Stack{op, dst} => format!("{:?} {:?}", op, dst),
            Instruction::Test{op, arg1, arg2} => format!("{:?} {:?} {:?}", op, arg1, arg2),
            Instruction::Flow{op, offset} => format!("{:?} {:?}", op, offset),
            Instruction::Other{op} => format!("{:?}", op)
        }
    }

    fn from_str(instruction_str : &str) -> Instruction{
        let args = instruction_str.split("").collect::<Vec<&str>>();
        // TODO
        // just so this will compile
        Instruction::BinArith{
            op: BinArithOp::ADD,
            dst: Register::R1,
            arg1: Register::R1,
            arg2: RegOrImm::Reg(Register::R2),  // TODO: refactor to RegOrImm::from method
        }
        
    }

}